phip/hit_calling.py

```diff
@@ -0,0 +1,285 @@
+import time
+
+import numpy as np
+import pandas as pd
+from scipy.optimize import minimize_scalar
+
+
+def do_hit_calling(
+        counts_df,
+        beads_only_samples,
+        fdr=0.15,
+        smoothing_bracket=(0, 500),
+        verbosity=2):
+    """
+    Call hits at the specified FDR using a heuristic.
+
+    A hit is a (clone, sample) pair whose read count exceeds what is expected
+    by chance.
+
+    The basic assumption is that the beads-only (negative control) samples
+    for some clone c give a good indication of the read count distribution
+    of non-hits for that clone. Background effects that are not captured by
+    beads-only samples will not be accounted for by this routine. At least
+    two beads-only samples are required, but more are better.
+
+    Overview of the algorithm:
+
+    An intuitive quantity to consider for a clone c and sample s is the
+    "smoothed ratio" between the read count at entry (c, s) and the maximum
```

Comment:
Why is the "maximum" the intuitive thing? That means the more bead-only controls you add, the more likely one of them will have a large fluctuation, the more difficult it is to get an outlier, no?

Reply:
It's a good point. The intuition is just that it wouldn't seem reasonable to me to call a hit for a sample/clone pair if some beads only sample for that clone had more reads (in our current regime of number of samples/number of beads). As we add more beads-only samples, the second-max will get closer to the max [1], which allows us to use a lower ratio threshold to achieve the same FDR. For this reason I don't think we'd hit the issue you mention, where adding more beads only samples means we call fewer hits.

On the other hand, you're right that in the limit of infinite beads-only samples, we won't ever call a hit because we limit our ratio threshold to be > 1. One possible fix there could be taking something like 99.99th percentile instead of the max. I'd say let's wait until we hit such a scenario though, as I doubt we will see that as a practical issue.

[1] This has been shown for the exponential distribution (see [here]( http://www.stat.purdue.edu/~dasgupta/orderstats.pdf), theorem 6.5). In that link, the distribution of the "spacing" between consecutive order statistics is derived, and as sample size increases this distribution gets more and more concentrated around 0. I'm not sure if it holds in general, although intuitively it feels like it should to me, at least for non pathological distributions.